File: back-end-nestjs/src/modules/result/result.service.ts
import {
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Result } from './entities/result.entity';
import { Response } from '../response/entities/response.entity';
import { UpsertResultDto } from './dto/upsert-result.dto';

@Injectable()
export class ResultService {
  constructor(
    @InjectRepository(Result) private readonly resultRepo: Repository<Result>,
    @InjectRepository(Response)
    private readonly responseRepo: Repository<Response>,
  ) {}

  async upsert(dto: UpsertResultDto) {
    const response = await this.responseRepo.findOne({
      where: { id: dto.response_id },
      relations: ['project'],
    });
    if (!response) {
      throw new NotFoundException('response.not_found');
    }
    if (response.project?.active === false) {
      throw new NotFoundException('project.not_found');
    }

    let entity = await this.resultRepo.findOne({
      where: { response: { id: response.id } },
      relations: ['response', 'project'],
    });

    if (!entity) {
      entity = this.resultRepo.create({
        response,
        project: response.project ?? null,
        summary: dto.summary,
      });
    } else {
      entity.summary = dto.summary;
      entity.project = response.project ?? null;
    }

    return this.resultRepo.save(entity);
  }

  async findByResponse(responseId: number) {
    const entity = await this.resultRepo.findOne({
      where: { response: { id: responseId } },
      relations: ['response', 'project'],
    });
    if (!entity || entity.project?.active === false) {
      throw new NotFoundException('result.not_found');
    }
    return entity;
  }
}

File: back-end-java/src/main/java/com/aie/backend/modules/result/service/ResultService.java
package com.aie.backend.modules.result.service;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.aie.backend.modules.project.entities.Project;
import com.aie.backend.modules.response.entities.Response;
import com.aie.backend.modules.response.repository.ResponseRepository;
import com.aie.backend.modules.result.dto.UpsertResultRequest;
import com.aie.backend.modules.result.entities.Result;
import com.aie.backend.modules.result.repository.ResultRepository;

@Service
@Transactional
public class ResultService {

    private final ResultRepository resultRepository;
    private final ResponseRepository responseRepository;

    public ResultService(ResultRepository resultRepository, ResponseRepository responseRepository) {
        this.resultRepository = resultRepository;
        this.responseRepository = responseRepository;
    }

    public Result upsert(UpsertResultRequest request) {
        Response response = responseRepository.findById(request.getResponseId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "response.not_found"));
        Project project = response.getProject();
        if (project != null && Boolean.FALSE.equals(project.getActive())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "project.not_found");
        }

        Result entity = resultRepository.findByResponseId(response.getId()).orElse(null);
        if (entity == null) {
            entity = new Result();
            entity.setResponse(response);
        }
        entity.setProject(response.getProject());
        entity.setSummary(request.getSummary());

        return resultRepository.save(entity);
    }

    @Transactional(readOnly = true)
    public Result findByResponse(Long responseId) {
        Result result = resultRepository.findByResponseId(responseId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "result.not_found"));
        if (result.getProject() != null && Boolean.FALSE.equals(result.getProject().getActive())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "result.not_found");
        }
        return result;
    }
}
